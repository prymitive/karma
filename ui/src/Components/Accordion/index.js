import React from "react";

import Collapsible from "react-collapsible";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronUp } from "@fortawesome/free-solid-svg-icons/faChevronUp";
import { faChevronDown } from "@fortawesome/free-solid-svg-icons/faChevronDown";

const Trigger = ({ text, isOpen }) => (
  <div className="d-flex flex-row justify-content-between">
    <div>{text}</div>
    <div>
      <FontAwesomeIcon
        icon={isOpen ? faChevronDown : faChevronUp}
        className="text-muted"
      />
    </div>
  </div>
);

const Accordion = ({ text, content, extraProps }) => (
  <Collapsible
    triggerTagName="div"
    transitionTime={50}
    className="card"
    openedClassName="card"
    triggerClassName="card-header cursor-pointer"
    triggerOpenedClassName="card-header cursor-pointer bg-light"
    contentOuterClassName="collapse show"
    contentInnerClassName="card-body my-2"
    {...extraProps}
    trigger={<Trigger text={text} isOpen={false} />}
    triggerWhenOpen={<Trigger text={text} isOpen={true} />}
  >
    {content}
  </Collapsible>
);

export { Accordion };
