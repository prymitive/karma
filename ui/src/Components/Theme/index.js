import React, { useEffect } from "react";
import ReactDOM from "react-dom";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSun } from "@fortawesome/free-solid-svg-icons/faSun";

const DarkTheme = React.lazy(() => import("Styles/DarkTheme"));
const LightTheme = React.lazy(() => import("Styles/LightTheme"));

const Placeholder = () => {
  return ReactDOM.createPortal(
    <div
      style={{
        zIndex: 2000,
        backgroundColor: "#455a64",
        position: "fixed",
        top: 0,
        left: 0,
        width: "100vw",
        height: "100vh",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        flexDirection: "column",
      }}
    >
      <FontAwesomeIcon
        icon={faSun}
        size="lg"
        spin
        style={{
          color: "#5e7a88",
          fontSize: "14rem",
        }}
      />
    </div>,
    document.body
  );
};

const ThemeContext = React.createContext();

const BodyTheme = () => {
  const context = React.useContext(ThemeContext);

  useEffect(() => {
    document.body.classList.toggle("theme-dark", context.isDark);
    document.body.classList.toggle("theme-light", !context.isDark);
  }, [context.isDark]);

  return (
    <React.Suspense fallback={<Placeholder />}>
      {context.isDark ? <DarkTheme /> : <LightTheme />}
    </React.Suspense>
  );
};

export { BodyTheme, ThemeContext };
