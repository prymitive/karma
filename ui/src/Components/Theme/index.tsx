import React, { FC, useEffect } from "react";
import ReactDOM from "react-dom";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSun } from "@fortawesome/free-solid-svg-icons/faSun";

import { Styles } from "react-select";

const DarkTheme = React.lazy(() => import("Styles/DarkTheme"));
const LightTheme = React.lazy(() => import("Styles/LightTheme"));

const Placeholder = () => {
  return ReactDOM.createPortal(
    <div
      style={{
        zIndex: 2000,
        backgroundColor: "#455a64",
        position: "fixed",
        top: 0,
        left: 0,
        width: "100vw",
        height: "100vh",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        flexDirection: "column",
      }}
    >
      <FontAwesomeIcon
        icon={faSun}
        size="lg"
        spin
        style={{
          color: "#5e7a88",
          fontSize: "14rem",
        }}
      />
    </div>,
    document.body
  );
};

export interface ThemeCtx {
  isDark: boolean;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  reactSelectStyles: Styles<any, any>;
  animations: {
    duration: number;
  };
}

const ThemeContext = React.createContext({
  isDark: false,
  reactSelectStyles: {},
  animations: { duration: 1000 },
} as ThemeCtx);

const BodyTheme: FC = () => {
  const context = React.useContext(ThemeContext);

  useEffect(() => {
    document.body.classList.toggle("theme-dark", context.isDark);
    document.body.classList.toggle("theme-light", !context.isDark);
  }, [context.isDark]);

  return (
    <React.Suspense fallback={<Placeholder />}>
      {context.isDark ? <DarkTheme /> : <LightTheme />}
    </React.Suspense>
  );
};

export { BodyTheme, ThemeContext };
