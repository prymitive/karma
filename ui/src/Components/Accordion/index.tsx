import React, { FunctionComponent, ReactNode, useState } from "react";

import { ToggleIcon } from "Components/ToggleIcon";

const Trigger: FunctionComponent<{ text: string; isOpen: boolean }> = ({
  text,
  isOpen,
}) => (
  <div className="d-flex flex-row justify-content-between">
    <div>{text}</div>
    <div>
      <ToggleIcon isOpen={isOpen} className="text-muted" />
    </div>
  </div>
);

const Accordion: FunctionComponent<{
  text: string;
  content: ReactNode;
  defaultIsOpen?: boolean;
}> = ({ text, content, defaultIsOpen }) => {
  const [isOpen, setIsOpen] = useState(defaultIsOpen || false);

  return (
    <div className="accordion card">
      <div
        className={`card-header cursor-pointer ${isOpen ? "active" : ""}`}
        onClick={() => setIsOpen(!isOpen)}
      >
        <Trigger text={text} isOpen={isOpen} />
      </div>
      <div className={isOpen ? "card-body my-2" : ""}>{isOpen && content}</div>
    </div>
  );
};

export { Accordion };
