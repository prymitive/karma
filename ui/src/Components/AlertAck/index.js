import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";

import { toJS } from "mobx";
import { useObserver } from "mobx-react-lite";

import addSeconds from "date-fns/addSeconds";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCheck } from "@fortawesome/free-solid-svg-icons/faCheck";
import { faCheckCircle } from "@fortawesome/free-solid-svg-icons/faCheckCircle";
import { faSpinner } from "@fortawesome/free-solid-svg-icons/faSpinner";
import { faExclamationCircle } from "@fortawesome/free-solid-svg-icons/faExclamationCircle";

import { APIGroup } from "Models/API";
import { AlertStore } from "Stores/AlertStore";
import {
  SilenceFormStore,
  MatchersFromGroup,
  GenerateAlertmanagerSilenceData,
} from "Stores/SilenceFormStore";
import { useFetchAny } from "Hooks/useFetchAny";
import { TooltipWrapper } from "Components/TooltipWrapper";

const AlertAck = ({ alertStore, silenceFormStore, group }) => {
  const [clusters, setClusters] = useState([]);
  const [upstreams, setUpstreams] = useState([]);
  const [currentCluster, setCurrentCluster] = useState(0);
  const [isAcking, setIsAcking] = useState(false);

  const { response, error, inProgress, reset } = useFetchAny(upstreams);

  const onACK = () => {
    setIsAcking(true);

    let author =
      silenceFormStore.data.author !== ""
        ? toJS(silenceFormStore.data.author)
        : toJS(alertStore.settings.values.alertAcknowledgement.author);

    if (alertStore.info.authentication.enabled) {
      silenceFormStore.data.author = toJS(
        alertStore.info.authentication.username
      );
      author = alertStore.info.authentication.username;
    }

    const alertmanagers = Object.entries(group.alertmanagerCount)
      .filter(([amName, alertCount]) => alertCount > 0)
      .map(([amName, _]) => amName);
    const clusters = Object.entries(
      alertStore.data.clustersWithoutReadOnly
    ).filter(([clusterName, clusterMembers]) =>
      alertmanagers.some((m) => clusterMembers.includes(m))
    );

    let c = [];
    for (const [clusterName, clusterMembers] of clusters) {
      const durationSeconds = toJS(
        alertStore.settings.values.alertAcknowledgement.durationSeconds
      );
      const commentPrefix = toJS(
        alertStore.settings.values.alertAcknowledgement.commentPrefix
      );
      const now = new Date();
      c.push({
        payload: GenerateAlertmanagerSilenceData(
          now,
          addSeconds(now, durationSeconds),
          MatchersFromGroup(group, [], group.alerts, true),
          author,
          `${
            commentPrefix ? commentPrefix + " " : ""
          }This alert was acknowledged using karma on ${now.toUTCString()}`
        ),
        clusterName: clusterName,
        members: clusterMembers,
      });
    }
    setClusters(c);
  };

  useEffect(() => {
    if (upstreams.length && !inProgress && (error || response)) {
      if (clusters.length > currentCluster + 1) {
        setCurrentCluster(currentCluster + 1);
      } else {
        setIsAcking(false);
      }
    }
  }, [clusters, upstreams, currentCluster, inProgress, error, response]);

  useEffect(() => {
    if (clusters.length) {
      reset();
      const cluster = clusters[currentCluster];
      let u = [];
      cluster.members.forEach((amName) => {
        const am = alertStore.data.getAlertmanagerByName(amName);
        if (am !== undefined) {
          u.push({
            uri: `${am.uri}/api/v2/silences`,
            options: {
              method: "POST",
              body: JSON.stringify(cluster.payload),
              credentials: am.corsCredentials,
              headers: {
                "Content-Type": "application/json",
                ...am.headers,
              },
            },
          });
        } else {
          console.error(`Alertmanager "${amName}" not found`);
        }
      });
      setUpstreams(u);
    }
  }, [alertStore.data, clusters, currentCluster, reset]);

  useEffect(() => {
    let timer;
    if (!isAcking && error) {
      timer = setTimeout(() => {
        setUpstreams([]);
        setIsAcking(false);
        reset();
      }, 20 * 1000);
    }
    return () => clearTimeout(timer);
  }, [isAcking, error, reset]);

  return useObserver(() =>
    alertStore.settings.values.alertAcknowledgement.enabled === false ? null : (
      <TooltipWrapper
        title={
          !isAcking && error
            ? error
            : "Acknowledge this alert with a short lived silence"
        }
      >
        <span
          className={`badge badge-pill components-label components-label-with-hover px-2 ${
            !isAcking && error
              ? "badge-warning"
              : !isAcking && response
              ? "badge-success"
              : "badge-secondary"
          }`}
          onClick={() => {
            if (!isAcking && !(response || error)) {
              setIsAcking(true);
              onACK();
            }
          }}
        >
          {!isAcking && error ? (
            <FontAwesomeIcon icon={faExclamationCircle} fixedWidth />
          ) : !isAcking && response ? (
            <FontAwesomeIcon icon={faCheckCircle} fixedWidth />
          ) : isAcking ? (
            <FontAwesomeIcon icon={faSpinner} fixedWidth spin />
          ) : (
            <FontAwesomeIcon icon={faCheck} fixedWidth />
          )}
        </span>
      </TooltipWrapper>
    )
  );
};
AlertAck.propTypes = {
  alertStore: PropTypes.instanceOf(AlertStore).isRequired,
  silenceFormStore: PropTypes.instanceOf(SilenceFormStore).isRequired,
  group: APIGroup.isRequired,
};

export { AlertAck };
