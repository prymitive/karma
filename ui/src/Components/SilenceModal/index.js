import React from "react";
import PropTypes from "prop-types";

import { useObserver } from "mobx-react-lite";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faBellSlash } from "@fortawesome/free-solid-svg-icons/faBellSlash";
import { faSpinner } from "@fortawesome/free-solid-svg-icons/faSpinner";

import { AlertStore } from "Stores/AlertStore";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { Settings } from "Stores/Settings";
import { Modal } from "Components/Modal";
import { TooltipWrapper } from "Components/TooltipWrapper";

// https://github.com/facebook/react/issues/14603
const SilenceModalContent = React.lazy(() =>
  import("./SilenceModalContent").then((module) => ({
    default: module.SilenceModalContent,
  }))
);

const SilenceModal = ({ alertStore, silenceFormStore, settingsStore }) => {
  // uses React.useCallback instead of useCallback for tests
  const onDeleteModalClose = React.useCallback(() => {
    const event = new CustomEvent("remountModal");
    window.dispatchEvent(event);
  }, []);

  return useObserver(() => (
    <React.Fragment>
      <li
        className={`nav-item components-navbar-button ${
          silenceFormStore.toggle.visible ? "border-info" : ""
        }`}
      >
        <TooltipWrapper title="New silence">
          <span
            id="components-new-silence"
            className="nav-link cursor-pointer"
            onClick={silenceFormStore.toggle.toggle}
          >
            <FontAwesomeIcon icon={faBellSlash} />
          </span>
        </TooltipWrapper>
      </li>
      <Modal
        isOpen={silenceFormStore.toggle.visible}
        toggleOpen={silenceFormStore.toggle.toggle}
        onExited={() => {
          silenceFormStore.data.resetProgress();
          silenceFormStore.data.autofillMatchers = true;
        }}
      >
        <React.Suspense
          fallback={
            <h1 className="display-1 text-placeholder p-5 m-auto">
              <FontAwesomeIcon icon={faSpinner} size="lg" spin />
            </h1>
          }
        >
          <SilenceModalContent
            alertStore={alertStore}
            silenceFormStore={silenceFormStore}
            settingsStore={settingsStore}
            onHide={silenceFormStore.toggle.hide}
            onDeleteModalClose={onDeleteModalClose}
          />
        </React.Suspense>
      </Modal>
    </React.Fragment>
  ));
};
SilenceModal.propTypes = {
  alertStore: PropTypes.instanceOf(AlertStore).isRequired,
  silenceFormStore: PropTypes.instanceOf(SilenceFormStore).isRequired,
  settingsStore: PropTypes.instanceOf(Settings).isRequired,
};

export { SilenceModal };
