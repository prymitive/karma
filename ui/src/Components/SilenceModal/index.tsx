import React, { FC } from "react";

import { observer } from "mobx-react-lite";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faBellSlash } from "@fortawesome/free-solid-svg-icons/faBellSlash";
import { faSpinner } from "@fortawesome/free-solid-svg-icons/faSpinner";

import { AlertStore } from "Stores/AlertStore";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { Settings } from "Stores/Settings";
import { Modal } from "Components/Modal";
import { TooltipWrapper } from "Components/TooltipWrapper";

// https://github.com/facebook/react/issues/14603
const SilenceModalContent = React.lazy(() =>
  import("./SilenceModalContent").then((module) => ({
    default: module.SilenceModalContent,
  }))
);

const SilenceModal: FC<{
  alertStore: AlertStore;
  silenceFormStore: SilenceFormStore;
  settingsStore: Settings;
}> = ({ alertStore, silenceFormStore, settingsStore }) => {
  return (
    <>
      <li
        className={`nav-item components-navbar-button ${
          silenceFormStore.toggle.visible ? "border-info" : ""
        } ml-auto`}
      >
        <TooltipWrapper title="New silence">
          <span
            id="components-new-silence"
            className="nav-link cursor-pointer"
            onClick={silenceFormStore.toggle.toggle}
          >
            <FontAwesomeIcon icon={faBellSlash} fixedWidth />
          </span>
        </TooltipWrapper>
      </li>
      <Modal
        isOpen={silenceFormStore.toggle.visible}
        toggleOpen={silenceFormStore.toggle.toggle}
        onExited={() => {
          silenceFormStore.data.resetProgress();
          silenceFormStore.data.setAutofillMatchers(true);
        }}
      >
        <React.Suspense
          fallback={
            <h1 className="display-1 text-placeholder p-5 m-auto">
              <FontAwesomeIcon icon={faSpinner} size="lg" spin />
            </h1>
          }
        >
          <SilenceModalContent
            alertStore={alertStore}
            silenceFormStore={silenceFormStore}
            settingsStore={settingsStore}
            onHide={silenceFormStore.toggle.hide}
          />
        </React.Suspense>
      </Modal>
    </>
  );
};

export default observer(SilenceModal);
