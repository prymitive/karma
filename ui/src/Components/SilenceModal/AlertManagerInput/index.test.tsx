import { mount } from "enzyme";

import toDiffableHtml from "diffable-html";

import { MockThemeContext } from "__fixtures__/Theme";
import { AlertStore } from "Stores/AlertStore";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { ThemeContext } from "Components/Theme";
import { AlertManagerInput } from ".";
import { APIAlertsResponseUpstreamsT } from "Models/APITypes";

let alertStore: AlertStore;
let silenceFormStore: SilenceFormStore;

const generateUpstreams = (): APIAlertsResponseUpstreamsT => ({
  counters: { total: 3, healthy: 3, failed: 0 },
  instances: [
    {
      name: "am1",
      uri: "http://am1.example.com",
      publicURI: "http://am1.example.com",
      readonly: false,
      headers: {},
      corsCredentials: "include",
      error: "",
      version: "0.17.0",
      cluster: "HA",
      clusterMembers: ["am1", "am2"],
    },
    {
      name: "am2",
      uri: "http://am2.example.com",
      publicURI: "http://am2.example.com",
      readonly: false,
      headers: {},
      corsCredentials: "include",
      error: "",
      version: "0.17.0",
      cluster: "HA",
      clusterMembers: ["am1", "am2"],
    },
    {
      name: "am3",
      uri: "http://am3.example.com",
      publicURI: "http://am3.example.com",
      readonly: false,
      headers: {},
      corsCredentials: "include",
      error: "",
      version: "0.17.0",
      cluster: "am3",
      clusterMembers: ["am3"],
    },
  ],
  clusters: { HA: ["am1", "am2"], am3: ["am3"] },
});

beforeEach(() => {
  alertStore = new AlertStore([]);
  alertStore.data.setUpstreams(generateUpstreams());
  silenceFormStore = new SilenceFormStore();
});

const MountedAlertManagerInput = () => {
  return mount(
    <AlertManagerInput
      alertStore={alertStore}
      silenceFormStore={silenceFormStore}
    />,
    {
      wrappingComponent: ThemeContext.Provider,
      wrappingComponentProps: { value: MockThemeContext },
    }
  );
};

const ValidateSuggestions = () => {
  const tree = MountedAlertManagerInput();
  // clear all selected instances, they are selected by default
  const clear = tree.find("ClearIndicator");
  // https://github.com/JedWatson/react-select/blob/c22d296d50917e210836fb011ae3e565895e6440/src/__tests__/Select.test.js#L1873
  clear.simulate("mousedown", { button: 0 });
  // click on the react-select component doesn't seem to trigger options
  // rendering in tests, so change the input instead
  tree.find("input").simulate("change", { target: { value: "am" } });
  return tree;
};

describe("<AlertManagerInput />", () => {
  it("matches snapshot", () => {
    const tree = MountedAlertManagerInput();
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("doesn't render ValidationError after passed validation", () => {
    const tree = MountedAlertManagerInput();
    silenceFormStore.data.setWasValidated(true);
    expect(toDiffableHtml(tree.html())).not.toMatch(/fa-exclamation-circle/);
    expect(toDiffableHtml(tree.html())).not.toMatch(/Required/);
  });

  it("renders ValidationError after failed validation", () => {
    const tree = MountedAlertManagerInput();
    tree.find(".react-select__multi-value__remove").at(0).simulate("click");
    tree.find(".react-select__multi-value__remove").at(0).simulate("click");
    silenceFormStore.data.setAlertmanagers([]);
    silenceFormStore.data.setWasValidated(true);
    expect(toDiffableHtml(tree.html())).toMatch(/fa-exclamation-circle/);
    expect(toDiffableHtml(tree.html())).toMatch(/Required/);
  });

  it("all available Alertmanager instances are selected by default", () => {
    MountedAlertManagerInput();
    expect(silenceFormStore.data.alertmanagers).toHaveLength(2);
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "Cluster: HA",
      value: ["am1", "am2"],
    });
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "am3",
      value: ["am3"],
    });
  });

  it("doesn't override last selected Alertmanager instances on mount", () => {
    silenceFormStore.data.setAlertmanagers([{ label: "am3", value: ["am3"] }]);
    MountedAlertManagerInput();
    expect(silenceFormStore.data.alertmanagers).toHaveLength(1);
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "am3",
      value: ["am3"],
    });
  });

  it("renders all 3 suggestions", () => {
    const tree = ValidateSuggestions();
    const options = tree.find("div.react-select__option");
    expect(options).toHaveLength(2);
    expect(options.at(0).text()).toBe("Cluster: HA");
    expect(options.at(1).text()).toBe("am3");
  });

  it("clicking on options appends them to silenceFormStore.data.alertmanagers", () => {
    silenceFormStore.data.setAlertmanagers([]);
    const tree = ValidateSuggestions();
    const options = tree.find("div.react-select__option");
    options.at(0).simulate("click");
    options.at(1).simulate("click");
    expect(silenceFormStore.data.alertmanagers).toHaveLength(2);
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "Cluster: HA",
      value: ["am1", "am2"],
    });
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "am3",
      value: ["am3"],
    });
  });

  it("silenceFormStore.data.alertmanagers gets updated from alertStore.data.upstreams.instances on mismatch", () => {
    MountedAlertManagerInput();
    alertStore.data.setClusters({
      amNew: ["amNew"],
    });
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "amNew",
      value: ["amNew"],
    });
  });

  it("is enabled when silenceFormStore.data.silenceID is null", () => {
    silenceFormStore.data.setSilenceID(null);
    const tree = MountedAlertManagerInput();
    const select = tree.find("StateManager");
    expect((select.props() as any).isDisabled).toBeFalsy();
  });

  it("is disabled when silenceFormStore.data.silenceID is not null", () => {
    silenceFormStore.data.setSilenceID("1234");
    const tree = MountedAlertManagerInput();
    const select = tree.find("StateManager");
    expect((select.props() as any).isDisabled).toBe(true);
  });

  it("removing last options sets silenceFormStore.data.alertmanagers to []", () => {
    const tree = MountedAlertManagerInput();
    expect(silenceFormStore.data.alertmanagers).toHaveLength(2);

    tree.find(".react-select__multi-value__remove").at(0).simulate("click");
    expect(silenceFormStore.data.alertmanagers).toHaveLength(1);

    tree.find(".react-select__multi-value__remove").simulate("click");
    expect(silenceFormStore.data.alertmanagers).toHaveLength(0);
    expect(silenceFormStore.data.alertmanagers).toEqual([]);
  });

  it("doesn't include readonly instances", () => {
    const upstreams = generateUpstreams();
    upstreams.instances[0].readonly = true;
    upstreams.instances[2].readonly = true;
    MountedAlertManagerInput();
    alertStore.data.setUpstreams(upstreams);
    expect(silenceFormStore.data.alertmanagers).toHaveLength(1);
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "am2",
      value: ["am2"],
    });
  });
});
