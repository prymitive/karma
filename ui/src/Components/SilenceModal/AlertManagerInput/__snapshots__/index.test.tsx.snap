// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlertManagerInput /> matches snapshot 1`] = `
"
<div class=\\" css-2b097c-container\\">
  <span aria-live=\\"polite\\"
        aria-atomic=\\"false\\"
        aria-relevant=\\"additions text\\"
        class=\\"css-1f43avz-a11yText-A11yText\\"
  >
  </span>
  <div class=\\"react-select__control css-cat782-control\\">
    <div class=\\"react-select__value-container react-select__value-container--is-multi react-select__value-container--has-value css-115r1xl-ValueContainer\\">
      <div class=\\"css-owt1hd-multiValue react-select__multi-value\\">
        <div class=\\"css-xbn6jz react-select__multi-value__label\\">
          Cluster: HA
        </div>
        <div class=\\"css-1ucqin4 react-select__multi-value__remove\\">
          <svg height=\\"14\\"
               width=\\"14\\"
               viewbox=\\"0 0 20 20\\"
               aria-hidden=\\"true\\"
               focusable=\\"false\\"
               class=\\"css-tj5bde-Svg\\"
          >
            <path d=\\"M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z\\">
            </path>
          </svg>
        </div>
      </div>
      <div class=\\"css-owt1hd-multiValue react-select__multi-value\\">
        <div class=\\"css-xbn6jz react-select__multi-value__label\\">
          am3
        </div>
        <div class=\\"css-1ucqin4 react-select__multi-value__remove\\">
          <svg height=\\"14\\"
               width=\\"14\\"
               viewbox=\\"0 0 20 20\\"
               aria-hidden=\\"true\\"
               focusable=\\"false\\"
               class=\\"css-tj5bde-Svg\\"
          >
            <path d=\\"M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z\\">
            </path>
          </svg>
        </div>
      </div>
      <div class=\\"css-ps6ina-Input\\">
        <div class=\\"react-select__input\\"
             style=\\"display: inline-block;\\"
        >
          <input autocapitalize=\\"none\\"
                 autocomplete=\\"off\\"
                 autocorrect=\\"off\\"
                 id=\\"react-select-silence-input-alertmanagers-input\\"
                 spellcheck=\\"false\\"
                 tabindex=\\"0\\"
                 type=\\"text\\"
                 aria-autocomplete=\\"list\\"
                 style=\\"box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;\\"
                 value
          >
          <div style=\\"position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;\\">
          </div>
        </div>
      </div>
    </div>
    <div class=\\"react-select__indicators css-vcwr3k-IndicatorsContainer\\">
      <div class=\\"react-select__indicator react-select__clear-indicator css-tlfecz-indicatorContainer\\"
           aria-hidden=\\"true\\"
      >
        <svg height=\\"20\\"
             width=\\"20\\"
             viewbox=\\"0 0 20 20\\"
             aria-hidden=\\"true\\"
             focusable=\\"false\\"
             class=\\"css-tj5bde-Svg\\"
        >
          <path d=\\"M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z\\">
          </path>
        </svg>
      </div>
      <span class=\\"react-select__indicator-separator css-1okebmr-indicatorSeparator\\">
      </span>
      <div class=\\"react-select__indicator react-select__dropdown-indicator css-tlfecz-indicatorContainer\\"
           aria-hidden=\\"true\\"
      >
        <svg height=\\"20\\"
             width=\\"20\\"
             viewbox=\\"0 0 20 20\\"
             aria-hidden=\\"true\\"
             focusable=\\"false\\"
             class=\\"css-tj5bde-Svg\\"
        >
          <path d=\\"M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z\\">
          </path>
        </svg>
      </div>
    </div>
  </div>
</div>
"
`;
